keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,832 @@
+    | status:                              |
+    |   conditions:                        |
+    |   - type: BucketReady                |
+    |     status: False                    |
+    |--------------------------------------|
+```
 
-The contents of the secret generated by the provisioner is expected to have the following fields in the provider specific format:
+###### 3. COSI calls appropriate driver to create Bucket
 
-- Endpoint for the bucket. The Endpoint should encode the URL for the provider
-- Access Key: Only set when serviceAccount is empty 
-- Secret Key: Only set when serviceAccount is empty
+The COSI sidecar, which runs alongside each of the drivers, listens for Bucket events. All but the sidecar with the specified provisioner will ignore the Bucket object. Only the appropriate sidecar will make a gRPC request to create a Bucket in the OSP.
 
-### Topology
+More information about COSI gRPC API is [here](#cosi-grpc-api)
 
-![Architecture Diagram](images/arch.png)
+```
+    COSI Driver (s3.amazonaws.com)
+    |------------------------------------------|
+    | grpc ProvisionerCreateBucket({           |
+    |      "name": "bcl-$uuid",                |
+    |      "protocol": "s3",                   |
+    |      "parameters": {                     |
+    |         "key": "value"                   |
+    |      }                                   |
+    | })                                       |
+    |------------------------------------------|
+```
+Once the Bucket is successfully created in the OSP, and a successful status is retreieved from the gRPC request, sidecar sets `BucketReady`condition to `True`. At this point, the Bucket is ready to be utilized by workloads.
+
+## Generating Access Credentials for Buckets
+
+The following stakeholders are involved in the lifecycle of access credential generation:
+
+ - Users  - request access to buckets
+ - Admins - establish cluster wide access policies
+
+Access credentials are represented by BucketAccess objects. Here are the steps for creating a BucketAccess:
+
+###### 1. Admin creates BucketAccessClass, User creates BucketAccess
 
-# Object Relationships
+The BucketAccessClass represents a set of common properties shared by multiple BucketClaims. It is used to specify policies for creating access credentials, and also for configuring driver-specific access parameters. More information about BucketAccessClass is [here](#bucketaccessclass)
 
-The diagram below describes the relationships between various resources in the COSI ecosystem. 
+The BucketAccess is used to request access to a bucket. It contains fields for choosing the Bucket for which the credentials will be generated, and also includes a bucketAccessClassName field, which in-turn contains configuration for authorizing users to access buckets. More information about BucketAccess is [here](#bucketaccess)
+
+BucketAccessClass can be used to specify a authorization mechanism. It can be one of 
+ - KEY  (__default__) 
+ - IAM 
```

Comment:
Since using IAM or KEY requires the application to be aware, should this be specified in the BucketClaim? That way a provisioner can reject the claim if it doesn't support it.

Reply:
If the field were moved to the BucketClaim, then it stops being portable when moving from a provider that supports IAM to another that doesn't support it. i.e. moving from S3 to MinIO. By making it a part of the class (an admin maintained object), the user resources are completely reusable even when porting to a different OSP.

Reply:
Similar to Protocol, since the application already has to be written with and aware of these attributes, I think it makes sense that they need to request it as a provisioning option. If they move to another environment with another storage provider, that storage provider needs to be able to support all the same features that the application needs.

If it's only in the storageclass, and you move to another provider that doesn't support that feature, then yes provisioning will succeed, but the application is still broken.

Reply:
Provisioning will not actually succeed, because the driver will reject the request if it doesn't support the requested feature.